### PR TITLE
Computed velocities and GRGS dialect

### PR DESCRIPTION
**Pull request by eggrobin**
Use the finite differences added in #2123.

### CODE REVIEW

File: astronomy/standard_product_3.cpp [hunk 1]
```diff
@@ -10,20 +10,72 @@
...
+using numerics::FiniteDifference;
 using quantities::NaN;
 using quantities::Speed;
 using quantities::si::Deci;
 using quantities::si::Kilo;
 using quantities::si::Metre;
 using quantities::si::Second;
 
+// Given a trajectory whose velocities are bad or absent (e.g., NaN), uses
+// n-point finite difference formulae on the positions to produce a trajectory
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Spelling!

File: astronomy/standard_product_3.cpp [hunk 2]
```diff
@@ -10,20 +10,72 @@
...
+             offset)});
+    // At every iteration, either |offset| advances, or the |positions|
+    // window shifts and |it| advances.
+    if (offset < (n - 1) / 2 || it == arc.End()) {
+      ++offset;
+    } else {
+      for (int k = 0; k < n - 1; ++k) {
+        times[k] = times[k + 1];
+        positions[k] = positions[k + 1];
+      }
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Use the 3-argument `std::move`.

File: astronomy/standard_product_3.cpp [hunk 3]
```diff
@@ -309,30 +366,76 @@ StandardProduct3::StandardProduct3(
...
+    if (orbit.back()->Empty()) {
+      orbit.pop_back();
+    }
+  }
   CHECK(!line.has_value()) << location;
+  if (!has_velocities_) {
+    for (auto& [id, orbit] : orbits_) {
+      for (auto& arc : orbit) {
+        switch (arc->Size()) {
+#define COMPUTE_VELOCITIES_CASE(n)    \
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Align the `\` (clang-format should do it).

Also, preference for *not* (un)defining macros in the middle of statements.

File: astronomy/standard_product_3.cpp [hunk 4]
```diff
@@ -309,30 +366,76 @@ StandardProduct3::StandardProduct3(
...
+          COMPUTE_VELOCITIES_CASE(1);
+          COMPUTE_VELOCITIES_CASE(2);
+          COMPUTE_VELOCITIES_CASE(3);
+          COMPUTE_VELOCITIES_CASE(4);
+          COMPUTE_VELOCITIES_CASE(5);
+          COMPUTE_VELOCITIES_CASE(6);
+          COMPUTE_VELOCITIES_CASE(7);
+          COMPUTE_VELOCITIES_CASE(8);
+#undef COMPUTE_VELOCITIES_CASE
+          default:
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: This is weird.  Why not a normal `9` branch and a `LOG(FATAL)` here?

File: astronomy/standard_product_3.hpp [hunk 5]
```diff
@@ -80,13 +93,22 @@ class StandardProduct3 {
   // (that order is the same in the satellite ID records and within each epoch).
   std::vector<SatelliteIdentifier> const& satellites() const;
 
-  DiscreteTrajectory<ITRS> const& orbit(SatelliteIdentifier const& id) const;
+  // Each orbit may consists of several arcs, separated by missing data.
```
Comment (pleroy) [inline astronomy/standard_product_3.hpp]: "consist"

Also mention that the trajectories are in time order (hopefully).

File: astronomy/standard_product_3_test.cpp [hunk 6]
```diff
@@ -2,26 +2,56 @@
...
 
 namespace principia {
 namespace astronomy {
 
+using base::dynamic_cast_not_null;
+using base::not_null;
+using geometry::Position;
+using physics::BodySurfaceDynamicFrame;
+using physics::ContinuousTrajectory;
+using physics::DiscreteTrajectory;
```
Comment (pleroy) [inline astronomy/standard_product_3_test.cpp]: Alphabetize.

File: astronomy/standard_product_3_test.cpp [hunk 7]
```diff
@@ -2,26 +2,56 @@
...
 using ::testing::Field;
+using ::testing::Lt;
 using ::testing::ResultOf;
 using ::testing::SizeIs;
 using ::testing::UnorderedElementsAre;
+using ::testing::ValuesIn;
 
 class StandardProduct3Test : public ::testing::Test {
  protected:
+  StandardProduct3Test() {}
```
Comment (pleroy) [inline astronomy/standard_product_3_test.cpp]: `= default`?  Why is this even needed?

File: astronomy/standard_product_3.cpp (hunk 4 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: There can be (and most often will be) more than 9 points in the orbit.
The template parameter, being the number of points used by the finite difference formula, must be smaller than or equal to the number of points in the orbit, but cannot be arbitrarily large (we only have coefficients for up to 9-point difference formulæ).

File: astronomy/standard_product_3_test.cpp (hunk 7 above)
Comment (eggrobin) [reply astronomy/standard_product_3_test.cpp]: It is not.

File: astronomy/standard_product_3_test.cpp (hunk 6 above)
Comment (eggrobin) [reply astronomy/standard_product_3_test.cpp]: Done.

File: astronomy/standard_product_3.hpp (hunk 5 above)
Comment (eggrobin) [reply astronomy/standard_product_3.hpp]: Done.

File: astronomy/standard_product_3.cpp (hunk 3 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: Done.

File: astronomy/standard_product_3.cpp (hunk 2 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: Done.

File: astronomy/standard_product_3.cpp (hunk 1 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: Done.

File: astronomy/geodesy_test.cpp [hunk 8]
```diff
@@ -106,19 +106,20 @@ TEST_F(GeodesyTest, LAGEOS2) {
   StandardProduct3::SatelliteIdentifier const lageos2_id{
       StandardProduct3::SatelliteGroup::General, 52};
 
-  CHECK_EQ(initial_ilrsa.orbit(lageos2_id).Begin().time(),
-           initial_ilrsb.orbit(lageos2_id).Begin().time());
+  CHECK_EQ(initial_ilrsa.orbit(lageos2_id)[0]->Begin().time(),
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply astronomy/geodesy_test.cpp]: Done.

**Pull request merged**